build-scripts/deploy-github.js

```diff
@@ -0,0 +1,176 @@
+#!/usr/bin/env node
+const fs = require('fs'),
+      os = require('os'),
+      request = require('request'),
+      uritemplate  = require('uri-template'),
+      path = require('path');
+
+const repoInfo = {
+    org: 'mozilla-magnet',
+    repo: 'magnet',
+};
+
+const userAgent = "mozmagnet Release-Agent";
+
+const githubAuthToken = process.env.GITHUB_AUTH_TOKEN;
+
+const rootUrl = getRepoUrl();
+
+function getRepoUrl() {
+    return `https://api.github.com/repos/${repoInfo.org}/${repoInfo.repo}`;
+}
+
+function getNewReleaseUrl() {
+    return `${rootUrl}/releases`
+}
+
+function getReleaseInfoUrl(tag) {
+    return `${rootUrl}/releases/tags/${tag}`;
+}
+
+function getUploadReleaseAssetUrl(release, binaryName) {
+    const template = uritemplate.parse(release.uploadUrl);
+    return template.expand({ name: binaryName, label: binaryName});
+}
+
+function getReleaseInfo() {
+    const artifact = process.env.ARTIFACT;
+    if (!artifact) {
+        throw new Error("No ARTIFACT environment variable");
+    }
+
+    const tag = process.env.TAG;
+    if (!tag) {
+        throw new Error("No TAG environment variable");
+    }
+
+    return {
+        binaryName: path.basename(artifact),
+        path: artifact,
+        tag: process.env.TAG
+    };
+}
+
+function uploadBinary() {
+  const releaseInfo = getReleaseInfo();
+
+  if(!fs.existsSync(releaseInfo.path)) {
+    throw new Error('Error reading file.');
+  }
+
+  tryGetReleaseForTag(releaseInfo.tag).then((release) => {
+    if (!release) {
+        return createRelease({
+              tag: releaseInfo.tag,
+              name: releaseInfo.tag,
+              isPreRelease: true,
+        });
+    }
+
+    return release;
+  }).then((release) => {
+      return uploadFileToRelease(release, releaseInfo.binaryName, releaseInfo.path);
+  }).catch((e) => {
+      setTimeout(function() {
+          throw e;
```

Comment:
Because: `Promise.reject(0).catch((e) => { throw e; })` results in a `Promise`, by throwing in a `setTimeout` we throw on a task run on the main event loop and we break out of the Promise chain.

It's a common implementation of `done` or `finally` for a `Promise` to avoid errors that are otherwise lost on the promise chain.

Not sure if there is an alternative way of doing this since I last looked tbf.